blockchain/blockchain.go

```diff
@@ -608,82 +657,219 @@ func (b *Blockchain) WriteHeadersWithBodies(headers []*types.Header) error {
+// - The trie roots match up (state, transactions, receipts, uncles)
+// - The receipts match up
+// - The execution result matches up
+func (b *Blockchain) verifyBlockBody(block *types.Block) error {
+	// Make sure the Uncles root matches up
 	if hash := buildroot.CalculateUncleRoot(block.Uncles); hash != block.Header.Sha3Uncles {
-		return fmt.Errorf(
+		b.logger.Error(fmt.Sprintf(
 			"uncle root hash mismatch: have %s, want %s",
 			hash,
 			block.Header.Sha3Uncles,
-		)
+		))
+
+		return ErrInvalidSha3Uncles
 	}
 
+	// Make sure the transactions root matches up
 	if hash := buildroot.CalculateTransactionsRoot(block.Transactions); hash != block.Header.TxRoot {
-		return fmt.Errorf(
+		b.logger.Error(fmt.Sprintf(
 			"transaction root hash mismatch: have %s, want %s",
 			hash,
 			block.Header.TxRoot,
-		)
+		))
+
+		return ErrInvalidTxRoot
+	}
+
+	// Execute the transactions in the block and grab the result
+	blockResult, executeErr := b.executeBlockTransactions(block)
+	if executeErr != nil {
+		return fmt.Errorf("unable to execute block transactions, %w", executeErr)
+	}
+
+	// Verify the local execution result with the proposed block data
+	if err := blockResult.verifyBlockResult(block); err != nil {
+		return fmt.Errorf("unable to verify block execution result, %w", err)
+	}
+
+	return nil
+}
+
+// verifyBlockResult verifies that the block transaction execution result
+// matches up to the expected values
+func (br *BlockResult) verifyBlockResult(referenceBlock *types.Block) error {
+	// Make sure the number of receipts matches the number of transactions
+	if len(br.Receipts) != len(referenceBlock.Transactions) {
+		return ErrInvalidReceiptsSize
+	}
+
+	// Make sure the world state root matches up
+	if br.Root != referenceBlock.Header.StateRoot {
+		return ErrInvalidStateRoot
+	}
+
+	// Make sure the gas used is valid
+	if br.TotalGas != referenceBlock.Header.GasUsed {
+		return ErrInvalidGasUsed
 	}
 
-	// Checks are passed, write the chain
+	// Make sure the receipts root matches up
+	receiptsRoot := buildroot.CalculateReceiptsRoot(br.Receipts)
+	if receiptsRoot != referenceBlock.Header.ReceiptsRoot {
+		return ErrInvalidReceiptsRoot
+	}
+
+	return nil
+}
+
+// executeBlockTransactions executes the transactions in the block locally,
+// and reports back the block execution result
+func (b *Blockchain) executeBlockTransactions(block *types.Block) (*BlockResult, error) {
 	header := block.Header
 
-	// Process and validate the block
-	res, err := b.processBlock(block)
+	parent, ok := b.readHeader(header.ParentHash)
+	if !ok {
+		return nil, ErrParentNotFound
+	}
+
+	blockCreator, err := b.consensus.GetBlockCreator(header)
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	txn, err := b.executor.ProcessBlock(parent.StateRoot, block, blockCreator)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := b.consensus.PreStateCommit(header, txn); err != nil {
+		return nil, err
 	}
 
+	_, root := txn.Commit()
```

Comment:
`executeBlockTransactions` is called from `verifyBlockBody`. If timeout happens and block proposer changes in IBFT, there will be time wrong state is applied for a while. Am I correct? I think will not affect transaction process but view function might return wrong result.